plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,391 @@
+
+    :param (str) direction: is the opposite of the direction of the line that
+        is draw in the HTML representation. It represents the direction that
+        will result from the two boxes resulting in the line dividing up an
+        HTML box in the preview of the dashboard.
+    :param (float) size: determins how big the first of the two boxes that
+        result in a split will be. This is in units of pixels.
+    """
+    is_horizontal = (direction == 'horizontal')
+    new_top_left_x = top_left_x + is_horizontal*0.5*box_w
+    new_top_left_y = top_left_y + (not is_horizontal)*0.5*box_h
+    new_box_w = (not is_horizontal)*box_w + is_horizontal
+    new_box_h = (not is_horizontal) + is_horizontal*box_h
+
+    html_box = """<!-- Draw some lines in -->
+          context.beginPath();
+          context.rect({top_left_x}, {top_left_y}, {box_w}, {box_h});
+          context.lineWidth = 1;
+          context.strokeStyle = 'black';
+          context.stroke();
+    """.format(top_left_x=new_top_left_x, top_left_y=new_top_left_y,
+               box_w=new_box_w, box_h=new_box_h)
+
+    index_for_new_box = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_for_new_box] + html_box +
+                      dashboard_html[index_for_new_box:])
+    return dashboard_html
+
+
+def add_html_text(dashboard_html, text, top_left_x, top_left_y, box_w, box_h):
+    """
+    Add a number to the middle of an HTML box.
+    """
+    html_text = """<!-- Insert box numbers -->
+          context.font = '{font_size}pt Times New Roman';
+          context.textAlign = 'center';
+          context.fillText({text}, {top_left_x} + 0.5*{box_w}, {top_left_y} + 0.5*{box_h});
+    """.format(text=text, top_left_x=top_left_x, top_left_y=top_left_y,
+               box_w=box_w, box_h=box_h, font_size=font_size)
+
+    index_to_add_text = dashboard_html.find('</script>') - 1
+    dashboard_html = (dashboard_html[:index_to_add_text] + html_text +
+                      dashboard_html[index_to_add_text:])
+    return dashboard_html
+
+
+class Dashboard(dict):
+    def __init__(self, content=None):
+        if content is None:
+            content = {}
+
+        self.box_ids_to_path = {}
+        if not content:
+            self['layout'] = _empty_box()
+            self['version'] = 2
+            self['settings'] = {}
+        else:
+            self['layout'] = content['layout']
+            self['version'] = content['version']
+            self['settings'] = content['settings']
+
+            self._assign_boxes_to_ids()
+
+    def _assign_boxes_to_ids(self):
+        self.box_ids_to_path = {}
+        all_nodes = []
+        node_gen = node_generator(self['layout'])
+
+        finished_iteration = False
+        while not finished_iteration:
+            try:
+                all_nodes.append(node_gen.next())
+            except StopIteration:
+                finished_iteration = True
```

Comment:
Yeah, so running `.next()` on the `node_gen` object will actually mutate, specifically remove a set of nodes. Eventually it will be empty and `node_gen` will result in an error, so in order to grab all the nodes and remember them, I store them all in a list `all_nodes`